firmware/tools/export_external_apps.py

```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2023 Bernd Herzog
+#
+# This file is part of PortaPack.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2, or (at your option)
+# any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; see the file COPYING.  If not, write to
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+
+import os
+import sys
+import struct
+import subprocess
+
+usage_message = """
+PortaPack external app image creator
+
+Usage: <command> <project source dir> <binary dir> <cmake objcopy path> <list of external image prefixes>
+"""
+
+if len(sys.argv) < 4:
+	print(usage_message)
+	sys.exit(-1)
+
+def read_image(path):
+	f = open(path, 'rb')
+	data = f.read()
+	f.close()
+	return data
+
+def write_image(data, path):
+	f = open(path, 'wb')
+	f.write(data)
+	f.close()
+
+def patch_image(image_data, search_address, replace_address):
+	if (len(image_data) % 4) != 0:
+		print("file size not divideable by 4")
+		sys.exit(-1)
+
+	external_application_image = bytearray()
+
+	for x in range(int(len(image_data)/4)):
+		snippet = image_data[x*4:4*(x+1)]
+		val = int.from_bytes(snippet, byteorder='little')
+
+		if val > search_address and (val - search_address) < (40*1024):
```

Comment:
Can you add a comment here? I'm not sure I understand where this magic 40K comes from and it would be helpful for future maintainers to understand the mapping process.

Reply:
added comment